File: src/lib/docs.ts
// PlotCraft 通用"固定分节文档集合"前端 wrapper（world/ 等目录）
//
// 数据约定（镜像 Rust `src-tauri/src/docs/mod.rs`，snake_case 跨 boundary）：
//   <project>/<collection>/<filename> —— 一节一个 md 文件（frontmatter 存 title/updated）
// invoke 参数顶层 key 传 camelCase（Tauri 自动转 snake_case，跟 lib/concept.ts 同款约定）

import { invoke } from '@tauri-apps/api/core'

import type { DocEntry } from '@/types/world'

/** 扫描项目 <collection>/ 全部分节，缺文件返回 exists=false + 空内容（不报错） */
export async function listDocs(projectRoot: string, collection: string): Promise<DocEntry[]> {
  return invoke<DocEntry[]>('list_docs', { projectRoot, collection })
}

/** 保存一节内容（atomic write + frontmatter title/updated）
 *  docId 不在 collection 注册表内 → 后端抛错 */
export async function saveDoc(
  projectRoot: string,
  collection: string,
  docId: string,
  content: string,
): Promise<DocEntry> {
  return invoke<DocEntry>('save_doc', { projectRoot, collection, docId, content })
}

/** 汇总 exists 且内容非空的分节（每节截断）—— 给 AI context 用
 *  全部为空 → 返回空串 */
export async function getDocsSummary(projectRoot: string, collection: string): Promise<string> {
  return invoke<string>('get_docs_summary', { projectRoot, collection })
}

File: src-tauri/src/commands/docs.rs
//! 固定分节文档集合（world/ 等目录）Tauri commands（v0.3）
//!
//! 3 个 command 全部 `async fn` + `AppResult<T>`，同步 fs 走 `spawn_blocking`
//! （照搬 commands/concept.rs 惯例）。写盘 atomic（tmp → rename，在 docs 模块内）。

use std::path::PathBuf;

use crate::docs::{
    docs_summary, save_doc as save_doc_impl, scan_docs, DocEntry, DEFAULT_SUMMARY_MAX_CHARS,
};
use crate::error::{AppError, AppResult};

/// 扫描项目某 collection 的固定分节，缺文件的分节返回 exists: false + 空内容
#[tauri::command]
pub async fn list_docs(project_root: String, collection: String) -> AppResult<Vec<DocEntry>> {
    let root = PathBuf::from(project_root);
    tokio::task::spawn_blocking(move || scan_docs(&root, &collection))
        .await
        .map_err(|e| AppError::Config(format!("list_docs: join: {}", e)))?
}

/// 保存一节（atomic write；懒建 collection 目录 —— 旧项目无目录也能直接 save）
/// frontmatter 只带 title + updated，不带 status
#[tauri::command]
pub async fn save_doc(
    project_root: String,
    collection: String,
    doc_id: String,
    content: String,
) -> AppResult<DocEntry> {
    let root = PathBuf::from(project_root);
    tokio::task::spawn_blocking(move || save_doc_impl(&root, &collection, &doc_id, &content))
        .await
        .map_err(|e| AppError::Config(format!("save_doc: join: {}", e)))?
}

/// 拼接 exists 且内容非空的分节成摘要（前端 AI context 用）
/// 全部缺文件/空内容 → 空串（前端据此跳过注入）
#[tauri::command]
pub async fn get_docs_summary(project_root: String, collection: String) -> AppResult<String> {
    let root = PathBuf::from(project_root);
    tokio::task::spawn_blocking(move || docs_summary(&root, &collection, DEFAULT_SUMMARY_MAX_CHARS))
        .await
        .map_err(|e| AppError::Config(format!("get_docs_summary: join: {}", e)))?
}
